### DISCUSSION TITLE
Warn about almost expired timeouts

### DISCUSSION THREAD

**Discussion by pohlt** [hynek/stamina#82] [Ideas]
If you specify a timeout and the action eventually worked but was close to timeout (e.g 80% of the timeout period passed), a warning would be helpful for the user. She/he could then increase the timeout or investigate the slow operation. The fraction of passed timeout for a warning could be configurable.

**Comment by hynek** [comment]
Not sure how that should work mechanically? You can approximate this by using the iterator-based approach and look at the [Attempt object](https://stamina.hynek.me/en/stable/api.html#stamina.Attempt)?

**Reply by pohlt** [reply]
Here's a quick hack of what I have in mind: https://github.com/hynek/stamina/compare/main...pohlt:stamina:main

Of course, the same would have to be done for coroutines, it needs to be configurable, etc.
With timeout it's a bit trickier since this relies on tenacity's functionality, but I could look into that as well if you are interested.
